include/SDL3/SDL_dialog.h

```diff
@@ -0,0 +1,190 @@
+
+  This software is provided 'as-is', without any express or implied
+  warranty.  In no event will the authors be held liable for any damages
+  arising from the use of this software.
+
+  Permission is granted to anyone to use this software for any purpose,
+  including commercial applications, and to alter it and redistribute it
+  freely, subject to the following restrictions:
+
+  1. The origin of this software must not be misrepresented; you must not
+     claim that you wrote the original software. If you use this software
+     in a product, an acknowledgment in the product documentation would be
+     appreciated but is not required.
+  2. Altered source versions must be plainly marked as such, and must not be
+     misrepresented as being the original software.
+  3. This notice may not be removed or altered from any source distribution.
+*/
+
+#ifndef SDL_dialog_h_
+#define SDL_dialog_h_
+
+#include <SDL3/SDL_video.h>
+
+#include <SDL3/SDL_begin_code.h>
+/* Set up for C function definitions, even when using C++ */
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * An entry for filters for file dialogs.
+ *
+ * `name` is a user-readable label for the filter (for example, "Office document").
+ *
+ * `pattern` is a semicolon-separated list of file extensions (for example,
+ * "doc;docx").
+ *
+ * \sa SDL_DialogFileCallback
+ * \sa SDL_ShowOpenFileDialog
+ * \sa SDL_ShowSaveFileDialog
+ * \sa SDL_ShowOpenFolderDialog
+ */
+typedef struct
+{
+    const char* name;
+    const char* pattern;
+} SDL_DialogFileFilter;
+
+/**
+ * Callback used by file dialog functions.
+ *
+ * The specific usage is described in each function.
+ *
+ * If @p filelist is...
+ * - `NULL`, an error occured. Details can be obtained with SDL_GetError().
+ * - A pointer to `NULL`, the user either didn't choose any file or canceled
+ *   the dialog.
+ * - A pointer to non-`NULL`, the user chose one or more files. The argument is
+ *   a null-terminated list of pointers to C strings, each containing a path.
+ *
+ * The @p filelist argument does not need to be freed; it will automatically be
+ * freed when the callback returns.
+ *
+ * The @p filter argument is the index of the filter that was selected, or one
+ * more than the size of the list (therefore the index of the terminating NULL
+ * entry) if no filter was selected, or -1 if the platform or method doesn't
+ * support fetching the selected filter or if an error occured.
+ *
+ * \sa SDL_DialogFileFilter
+ * \sa SDL_ShowOpenFileDialog
+ * \sa SDL_ShowSaveFileDialog
+ * \sa SDL_ShowOpenFolderDialog
+ */
+typedef void(*SDL_DialogFileCallback)(void* userdata, const char* const* filelist, int filter);
+
+/**
+ * Displays a dialog that lets the user select a file on their filesystem.
+ *
+ * This function should only be invoked from the main thread.
+ *
+ * This is an asynchronous function; it will return immediately, and the result
+ * will be passed to the callback.
```

Comment:
If this function is asynchronous, why should it only be called from the main thread?

Reply:
In short, Apple.

I don't remember the details very well, but macOS requires some GUI-related functions to be called from the main threads, I believe to avoid race conditions in the system libraries. And the system functions are themselves async, hence the strange pair of requirements.

I talked about it a while ago in [this comment](https://github.com/libsdl-org/SDL/pull/7770#issuecomment-1581309074), but the details are a bit fuzzy in my head.

Reply:
It might be possible to forward the open dialog request to the main thread's runloop on Apple platforms, I think some other SDL code does a similar thing. The callback would be called in the main thread even when the original function was not though, in that situation.

I'm not sure if other platforms would need similar forwarding to keep everything thread-safe. I don't think the external API would need to change if that threading restriction is relaxed in the future, so maybe it could be a post 3.0 TODO to investigate it?

Reply:
Okay, we can leave this as-is.